{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "typora-copy-images-to: ./\n",
    "sequence-fold: 80 xclip -o | tr -d \"\\n\" | fold -w80\n",
    "command-fold: 75 xclip -o | tr -d \"\\n\" | fold -sw75 | sed -e 's/$/\\\\/'\n",
    "last-update: April 2nd, 2018\n",
    "---\n",
    "\n",
    "# Introduction to BLAST \n",
    "\n",
    "by *alper yilmaz* for *GTU Bioinformatics Program Course*\n",
    "\n",
    "*2019-03-14* (PDF version of this document is accessible at `goo.gl/bhrkqQ` and html version is available at `https://goo.gl/siUoax`)\n",
    "\n",
    "<!-- long url is https://s3-us-west-2.amazonaws.com/veri-analizi/BLAST-outline.pdf -->\n",
    "\n",
    "<!-- TODO : html version need dataurlify utility -->\n",
    "\n",
    "[TOC]\n",
    "\n",
    "\n",
    "## Paralog vs. Homolog\n",
    "\n",
    "![Homology-Ortholog-Paralog](images/Homology.png)\n",
    "\n",
    "[image source](https://bitesizebio.com/26762/homology-terminology-never-say-wrong-word/)\n",
    "\n",
    "## Local vs. Global Alignment\n",
    "\n",
    "\n",
    "| Global Sequence Alignment                                    | Local  Sequence Alignment                                    |\n",
    "| :----------------------------------------------------------- | ------------------------------------------------------------ |\n",
    "| In global alignment, an attempt is made to align the entire sequence (end to end alignment) | Finds local regions with the highest level of similarity between the two sequences. |\n",
    "| A global alignment contains all letters from both the query and target sequences | A local alignment aligns a substring of the query sequence to a substring of the target sequence. |\n",
    "| If two sequences have approximately the same length and are quite similar, they are suitable for global alignment. | Any two sequences can be locally aligned as local alignment finds  stretches of sequences with high level of matches without  considering the alignment of rest of the sequence regions. |\n",
    "| Suitable for aligning two closely related sequences.         | Suitable for aligning more divergent sequences or distantly related sequences. |\n",
    "| Global alignments are usually done for comparing homologous genes like comparing two genes with same function (in human vs. mouse) or comparing two proteins with similar function. | Used for finding out conserved patterns in DNA sequences or conserved domains or motifs in two proteins. |\n",
    "| A general global alignment technique is the Needleman–Wunsch algorithm. | A general local alignment method is Smith–Waterman algorithm. |\n",
    "\n",
    "\n",
    "![Local vs Global Alignment](images/local_vs_global_alignment.jpg)\n",
    "\n",
    "\n",
    "[image and table source](https://www.majordifferences.com/2016/05/difference-between-global-and-local.html)\n",
    "\n",
    "## BLAST\n",
    "\n",
    "* **B**asic **L**ocal **A**lignment **S**earch **T**ool\n",
    "* http://blast.ncbi.nlm.nih.gov/\n",
    "* Are there sequences in databases/databanks similar to my sequence?\n",
    "\n",
    "**Functions of BLAST**\n",
    "\n",
    "* Identify species\n",
    "* Locating domains\n",
    "* Establishing phylogeny\n",
    "* Comparison\n",
    "\n",
    "## The Index\n",
    "\n",
    "![BLAST Index](images/ncbi-db.png)\n",
    "\n",
    "## Types of BLAST programs\n",
    "\n",
    "![BLAST programs](images/I.7_2_blast_types.png)\n",
    "\n",
    "[image source](http://library.open.oregonstate.edu/computationalbiology/chapter/command-line-blast/)\n",
    "\n",
    "## How BLAST works\n",
    "\n",
    "![How BLAST works](images/howblastworks.gif)\n",
    "\n",
    "[image source](https://www.integratedbreeding.net/courses/genomics-and-comparative-genomics/www.generationcp.org/genomics/index257b.html?page=1174)\n",
    "\n",
    "![BLAST mechanism](images/image011.gif)\n",
    "\n",
    "[image source](http://130.88.97.239/ember/prototype/CHAPTER02/INFORMATION.shtml)\n",
    "\n",
    "## Scoring\n",
    "\n",
    "### Nucleotide\n",
    "\n",
    "![Nucleotide scoring](images/scoring_nucleotides.gif)\n",
    "\n",
    "[image source](https://www.ncbi.nlm.nih.gov/Class/MLACourse/Modules/BLAST/scoring_nucleotides.html)\n",
    "\n",
    "### Alignment score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AACGTTTCCAGTCCAAATAGCTAGGC\n",
    "===--===   =-===-==-======\n",
    "AACCGTTC   TACAATTACCTAGGC\n",
    "\n",
    "Hits(+1): 18\n",
    "Misses (-2): 5\n",
    "Gaps (existence -2, extension -1): 1 Length: 3\n",
    "Score = 18 * 1 + 5 * (-2) – 2 – 2 = 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Amino Acid\n",
    "\n",
    "BLOSUM62 Substitution Matrix\n",
    "\n",
    "![BLOSUM62](images/BLOSUM.jpg)\n",
    "\n",
    "[image source](https://memim.com/blosum.html)"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
